### ISSUE TITLE
Crash of Leiden algorithm on 600 million vertices and 3.8 billion edge graphs.

### ISSUE THREAD

**Issue by munhyunsu** [igraph/igraph#1435]
I'm trying to run the Leiden algorithm on 600 million vertices and 3.8 billion edge graphs.
The graph is constructed with `read_graph_edge_list()` and the leiden algorithm is executed. The following message is displayed and a segmentation fault occurs.

I don't know if iGraph can't handle this graph or if there is an issue with the implementation of the leiden algorithm.
However, if the iGraph developer advises how to compile libigraph.so.0 in debug mode, as in the backtrace below, to find the location of the error (debug mode compilation method), I might contribute myself.

summary:
-Leiden algorithm fails on large graphs (600 million vertices, 3.8 billion Edge)

Backtrace:
```
#0 0x00007ffff786480c in igraph_i_community_leiden ()
   from /opt/igraph/lib/libigraph.so.0
#1 0x00007ffff78668ed in igraph_community_leiden ()
   from /opt/igraph/lib/libigraph.so.0
#2 0x0000555555554ee3 in main (argc=3, argv=0x7fffffffe0b8)
    at find_community_leiden.c:40
```

**Comment by vtraag** [comment]
Thanks for the report. Which version of `igraph` are you running?

**Comment by munhyunsu** [comment]
Hi @vtraag 
I used igraph-0.8.2

Thanks.

**Comment by vtraag** [comment]
Can you share the graph in one way or another? It is difficult to debug the problem without more detailed information.

You can also compile the `igraph` library yourself and enable debugging, preferably from the `master` branch of this repository. See https://igraph.org/c/#cinstall for the general instructions. Essentially, make sure you have installed the necessary packages. Assuming you are on Ubuntu or similar, simply run:
```
sudo apt-get install build-essential libxml2-dev libgmp-dev automake autoconf
```
after that, you should then run
```
./configure --enable-debug
make
sudo make install
```
You can also run `make check` to test the package. If you are indeed compiling from the `master` branch, you additionally need to install `flex` and `bison` and run `./bootstrap.sh` before running `./configure`.

**Comment by munhyunsu** [comment]
I was already compiling and using iGraph.
I will reinstall `igraph` with the `--enable-debug` tag to find the problem.

If you are interested in this data, please let me know.
Edge data is about 70 gigabytes, so I'll upload it to Google Drive and email the link to you.

**Comment by vtraag** [comment]
>  I will reinstall `igraph` with the `--enable-debug` tag to find the problem.

Great, hopefully that will provide some more insight!

> If you are interested in this data, please let me know.
> Edge data is about 70 gigabytes, so I'll upload it to Google Drive and email the link to you.

It might still be useful for debugging purposes. Although the question is: does the problem occur almost immediately after running, or does it take some time before the problem appears?

How much internal memory do you have? It could be that there is some out-of-memory error somewhere that is not caught for some reason.

**Comment by munhyunsu** [comment]
There is no error in the step of constructing the graph.
My code is [here](https://github.com/munhyunsu/BitcoinAnalysis/blob/master/BitcoinBlockSampler/find_community_leiden.c)

It works until you construct a graph in this code and initialize variables to find the community.
An error occurs in the middle of executing the leiden algorithm.

I have 512G memory.
(Actually, I am trying with C because iGraph-Python did not generate this graph.)

**Comment by vtraag** [comment]
>  An error occurs in the middle of executing the leiden algorithm.

OK, thanks for the clarification. My question is more about time: does it take 5 hours to crash, or does it crash almost immediately after starting to run `igraph_community_leiden`?

>  (Actually, I am trying with C because iGraph-Python did not generate this graph.)

That might be worth a separate issue. If you are using `Read_Edgelist` in Python, it essentially uses the same underlying C code.

**Comment by szhorvat** [comment]
As @vtraag suggested above, this is likely to be an out-of-memory error. While igraph does generally try to check for this situation, it is not common to run out of memory, so I would not be surprised if some checks are missing or incorrect.

If it really is an out-of-memory error, perhaps it can be triggered with a generated graph. Still, it's painful, because the first thing that happens with we run out of physical memory is that the OS starts swapping and everything becomes painfully slow. It won't fail immediately.

@munhyunsu Could you perhaps try to reproduce the problem with a generated graph having a similar number of edges and vertices as your actual dataset?

**Comment by ntamas** [comment]
>
> If it really is an out-of-memory error, perhaps it can be triggered with a
> generated graph. Still, it's painful, because the first thing that happens
> with we run out of physical memory is that the OS starts swapping and
> everything becomes painfully slow. It won't fail immediately.
>
I was wondering whether we could get around this by using a Docker
container with a limited memory allowance.

>

**Comment by vtraag** [comment]
> I was wondering whether we could get around this by using a Docker
> container with a limited memory allowance.

Yes, that is a good idea! I'll try to find some time to test that.

**Comment by vtraag** [comment]
I've been looking at using docker to test for memory issues, but it seems difficult to debug. I am running a docker container as follows
```bash
docker run --rm --memory 50m --memory-swap 50m -it -P --name test-igraph -v /home/vtraag/docker/data:/data ubuntu
```
Unfortunately, I'm just getting the following result

```bash
root@e9d97109e601:/data# ./igraphtest 
Created graph.
Running Leiden algorithm.
Killed
```

where `igraphtest` is a straightforward test:

```C
#include <igraph.h>

int main() {

    igraph_t g;
    igraph_vector_t membership, degree;
    igraph_integer_t n = 100000, nb_clusters;
    igraph_real_t quality;
    igraph_erdos_renyi_game(&g, IGRAPH_ERDOS_RENYI_GNM, n, 5*n, IGRAPH_UNDIRECTED, IGRAPH_NO_LOOPS);
    printf("Created graph.\n");
    igraph_vector_init(&membership, n);
    igraph_vector_init(&degree, n);

    igraph_degree(&g, &degree, igraph_vss_all(), IGRAPH_ALL, IGRAPH_NO_LOOPS);

    printf("Running Leiden algorithm.\n");
    igraph_community_leiden(&g, NULL, &degree, 1.0/(2*igraph_ecount(&g)), 0.01,
                            /*start=*/ 0, &membership, &nb_clusters, &quality);
    printf("Finished Leiden algorithm.\n");
    return 0;
}
```

The problem now is that the process is simply killed. Also if I try to run `gdb ./igraphtest` (first you need to create a docker container with `gdb` and run with `--security-opt seccomp=unconfined`), I get 
```bash
Reading symbols from ./igraphtest...
(gdb) run
Starting program: /data/igraphtest 
Created graph.
Running Leiden algorithm.

Program terminated with signal SIGKILL, Killed.
The program no longer exists.
```

Finally, there is the option to set `--oom-kill-disable` when running the docker container. Although the program is then not killed, it simply doesn't run to completion.

In short: I don't know how the trigger an out-of-memory error that will be caught while running `igraph` before the process is killed by the system. Any suggestions?

**Comment by ntamas** [comment]
Okay, I'm throwing a few more ideas around, not sure if any of these would
work, but these are the other options that I am aware of:

1. On Linux, you can use [cgroups](
https://man7.org/linux/man-pages/man7/cgroups.7.html) to control and limit
the memory usage of individual processes or process groups, irrespectively
of the total RAM available. So, you could take a test executable, put it
inside a control group with a limited amount of RAM available, and then
make it crash. The problem is that the OOM killer might simply send SIGKILL
to the process instead of returning NULL to the offending malloc() request.

2. Spin up a VM with VirtualBox and limit the amount of memory available to
the VM. Install Ubuntu, install igraph, disable the swap entirely and then
try to crash it. I think this is the closest that one can do to reproducing
an out-of-memory problem.

**Comment by vtraag** [comment]
I tried to replicate the problem with a VirtualBox, because, as you said @ntamas, that is closes to replicating a real out-of-memory problem. After limiting the swap and disabling the oom-killer (using `sysctl vm.overcommit_memory = 2`), I could finally start seeing some actual memory errors thrown by `igraph`, instead of the process simply being killed.

I was able to identify one problem. If `NDEBUG` is defined, `assert` statements will not be executed, and this might be a problem in some places. There are several places in which it is checked whether the vector has properly memory allocated (`assert(v->stor_begin != NULL);`) and other such checks. This is of course not done when `NDEBUG` is defined, in which case the error handling might not work exactly as intended. In that case, some problems may arise without clear error messages and without clean exits.

One example of this problem that I ran into was:
https://github.com/igraph/igraph/blob/a2e77aafdc2abcea42eec3b166f44f0988f31b2d/src/adjlist.c#L660-L662
The comment is only correct when `assert` is being run, otherwise, it might incorrectly try to free up a non-existing vector.

```c
Starting program: /media/share/igraphtest 

Program received signal SIGSEGV, Segmentation fault.
0x000000000047c572 in igraph_vector_int_destroy (v=0x0) at vector.pmt:380
380	vector.pmt: No such file or directory.
#0  0x000000000047c572 in igraph_vector_int_destroy (v=0x0) at vector.pmt:380
#1  0x0000000000488c17 in igraph_inclist_destroy (il=0x7fffffffe0f0) at adjlist.c:662
#2  0x000000000044de59 in IGRAPH_FINALLY_FREE () at igraph_error.c:207
#3  0x000000000044dccc in igraph_error_handler_ignore (
    reason=0x5dc680 "cannot reserve space for vector", file=0x5dc660 "vector.pmt", line=446, 
    igraph_errno=2) at igraph_error.c:159
#4  0x000000000044dae6 in igraph_error (reason=0x5dc680 "cannot reserve space for vector", 
    file=0x5dc660 "vector.pmt", line=446, igraph_errno=2) at igraph_error.c:118
#5  0x0000000000469d48 in igraph_vector_reserve (v=0x7fffffffdd80, size=463561) at vector.pmt:446
#6  0x000000000046a482 in igraph_vector_resize (v=0x7fffffffdd80, newsize=463561) at vector.pmt:897
#7  0x000000000048538f in igraph_vector_order (v=0x27e9b18, v2=0x27e9b30, res=0x7fffffffdd80, 
    nodes=23950) at vector.c:152
#8  0x00000000004455ec in igraph_add_edges (graph=0x27e9b10, edges=0x7fffffffdf10, attr=0x0)
    at type_indexededgelist.c:298
#9  0x0000000000439eb0 in igraph_create (graph=0x27e9b10, edges=0x7fffffffdf10, n=23950, 
    directed=0) at structure_generators.c:96
#10 0x000000000048f7db in igraph_i_community_leiden_aggregate (graph=0x2c5a750, 
    edges_per_node=0x7fffffffe0f0, edge_weights=0x292dca0, node_weights=0x2b16cc0, 
    membership=0x2aa4c80, refined_membership=0x7fffffffe160, nb_refined_clusters=23950, 
    aggregated_graph=0x27e9b10, aggregated_edge_weights=0x7fffffffe100, 
    aggregated_node_weights=0x7fffffffe120, aggregated_membership=0x7fffffffe140)
    at community_leiden.c:613
#11 0x000000000049060e in igraph_i_community_leiden (graph=0x7fffffffe2d0, edge_weights=0x675800, 
    node_weights=0x7fffffffe2b0, resolution_parameter=9.0909090909090904e-07, beta=0.01, 
    membership=0x7fffffffe290, nb_clusters=0x7fffffffe280, quality=0x7fffffffe288)
    at community_leiden.c:828
#12 0x0000000000490ed6 in igraph_community_leiden (graph=0x7fffffffe2d0, edge_weights=0x0, 
    node_weights=0x7fffffffe2b0, resolution_parameter=9.0909090909090904e-07, beta=0.01, start=0, 
    membership=0x7fffffffe290, nb_clusters=0x7fffffffe280, quality=0x7fffffffe288)
    at community_leiden.c:1069
#13 0x0000000000402605 in main () at src/test.c:17
quit
```

This is just one particular problem (which we should fix, obviously). However, we should probably look into this more generally. How are things typically compiled when included in package managers? It might make sense that they are using `NDEBUG`, in which case some problems might occur.

Now that I am thinking of this: it would (theoretically) be possible that `IGRAPH_ERROR` could be called while running `IGRAPH_FINALLY_FREE`, which could (theoretically) result in an infinite loop. Might it be an idea to guard against this in some way by settings a flag that indicates we are already running `IGRAPH_FINALLY_FREE`?

Now the question of course is: @munhyunsu did you compile defining `NDEBUG`? If not, I don't know what else I could do to replicate the problem, and I then hope you have been able to compile it using debug information and provide some more detail.

**Comment by munhyunsu** [comment]
Sorry for the late reply.
I am re-installing in `--enable-debug` mode to run the algorithm with the data below.
The results from the out-of-memory above are similar to my symptoms, but I'm not sure yet.
I will look at the trace in debug mode.

> did you compile defining NDEBUG?

Is `NDEBUG` a `--enable-debug`?

> Can you share the graph in one way or another? 

And I sent Edge data by email.

**Comment by vtraag** [comment]
> I will look at the trace in debug mode.

Great, looking forward to seeing more details!

> Is `NDEBUG` a `--enable-debug`?

No, it has to be set separately. Unless it is triggered by some other settings on your system, this is then unlikely to be the cause of your problem.

What OS and what version are you on?

**Comment by munhyunsu** [comment]
> What OS and what version are you on?

Linux Mint 19.3 (Dist based on Ubuntu 18.04) 64bit.

**Comment by szhorvat** [comment]
@vtraag Do you usually compile igraph with `NDEBUG`? I don't think it's a good idea to do that. I do not expect the `assert`s that are present to reduce performance, but when they do get triggered, they can produce very useful information.

This is why I always argue against using NDEBUG as a general debug flag, and coupling our own debug constructs to it: it is useful to leave the `assert`s in for release versions, but we don't want to have our own debug features enabled in release versions.

**Comment by munhyunsu** [comment]
> My question is more about time: does it take 5 hours to crash, or does it crash almost immediately after starting to run igraph_community_leiden?

I measured the time from graph generation to Leiden algorithm execution.
- Graph generation `igraph_read_graph_edgelist()`: 2023 seconds (33 minutes)
- From algorithm start `igraph_community_leiden` to error `Segmentation fault (core dumped)` : 152 seconds

```bash
$ inxi
CPU: 18-Core Intel Xeon Gold 5220 (-MT MCP-) 
speed/min/max: 1000/1000/3900 MHz Kernel: 5.4.0-40-generic x86_64 Up: 1h 31m 
Mem: 446.5/514452.1 MiB (0.1%) Storage: 2.96 TiB (48.4% used) Procs: 423 
Shell: bash 4.4.20 inxi: 3.0.32
```

**Comment by vtraag** [comment]
> Do you usually compile igraph with `NDEBUG`?

No, I have never done that, but is was the only thing that I have come across. I am not sure what are default flags that are used for compilation in different systems. If using `NDEBUG` results in a build that is incorrect, that should be clarified, and preferably it should be ensured that the build is still correct even if `NDEBUG` is defined.

I noticed some other potential issues, I will look into those next.

**Comment by munhyunsu** [comment]
@vtraag 
Unfortunately, the https://github.com/vtraag/igraph/tree/fix/1435 branch mentioned https://github.com/igraph/igraph/pull/1438 also fails.
When iGraph is installed with `--enable-debug` option and executed with gdb, the following message is displayed.

```
Program received signal SIGSEGV, Segmentation fault.
0x00007ffff7761d84 in igraph_i_community_leiden_fastmovenodes (
    graph=0x7fffffffdf10, edges_per_node=0x7fffffffdbc0,
    edge_weights=0x55555576a440, node_weights=0x7fffffffdef0,
    resolution_parameter=0.050000000000000003, nb_clusters=0x7fffffffde8c,
    membership=0x7fffffffded0) at community_leiden.c:145
145                 VECTOR(edge_weights_per_cluster)[c] += VECTOR(*edge_weights)[e];
```
```
(gdb) backtrace
#0  0x00007ffff7761d84 in igraph_i_community_leiden_fastmovenodes (
    graph=0x7fffffffdf10, edges_per_node=0x7fffffffdbc0, 
    edge_weights=0x55555576a440, node_weights=0x7fffffffdef0, 
    resolution_parameter=0.050000000000000003, nb_clusters=0x7fffffffde8c, 
    membership=0x7fffffffded0) at community_leiden.c:145
#1  0x00007ffff7764d90 in igraph_i_community_leiden (graph=0x7fffffffdf10, 
    edge_weights=0x55555576a440, node_weights=0x7fffffffdef0, 
    resolution_parameter=0.050000000000000003, beta=0.01, 
    membership=0x7fffffffded0, nb_clusters=0x7fffffffde8c, 
    quality=0x7fffffffdea8) at community_leiden.c:789
#2  0x00007ffff7765830 in igraph_community_leiden (graph=0x7fffffffdf10, 
    edge_weights=0x0, node_weights=0x7fffffffdef0, 
    resolution_parameter=0.050000000000000003, beta=0.01, start=0, 
    membership=0x7fffffffded0, nb_clusters=0x7fffffffde8c, 
    quality=0x7fffffffdea8) at community_leiden.c:1041
#3  0x0000555555554ee3 in main (argc=3, argv=0x7fffffffe0b8)
    at find_community_leiden.c:40
```

Could an Integer overflow be affected?
After creating a graph with `igraph_read_graph_edgelist()`, a negative number is output when printing the edge size.

```c
igraph_t graph;
FILE *input;
char *path = argv[1];
input = fopen(path, "r");
igraph_read_graph_edgelist(&graph, input, 0, 0);
printf("Graph load complete with %li vertices and %li edges\n",
        igraph_vcount(&graph),
        igraph_ecount(&graph));
```
```
Graph load complete with 628875886 vertices and -564832027 edges
```

Thanks for helping.

**Comment by vtraag** [comment]
@munhyunsu, recently we started work to shift towards using 64-bit integers. This functionality should be operational in the current `develop` branch of the C core. It has not yet fully propagated to the Python interface yet, which is still work in progress.

It would be useful to us if you would be able to help us out to test the changes in the C core. Would you be able to clone the `develop` branch, compile it, and run it on your data? If you need any help doing this, let us know.

**Comment by iosonofabio** [comment]
I think the develop branch of the Python interface had the 64 bit PR so perhaps worth a shot!
Fabio

**Comment by ntamas** [comment]
Yes, the `develop` branch of the Python interface should already have the
64-bit C core, but the Python-to-C glue layer might not be fully aware of
the changes. (We use long ints in lots of places where we should use an
igraph_integer_t instead). So it might not be working yet, but it's worth a
shot -- we will surely try to massage this until it works.

T.

**Comment by szhorvat** [comment]
With support for 64-bit integers, and overflow checks present in many function, this is likely fixed, so closing.

@munhyunsu If you can test it with [your short C program](https://github.com/igraph/igraph/issues/1435#issuecomment-664058584) and the `develop` branch of this repository, that would be quite useful though. Now igraph can be compiled with either 64-bit (default on 64-bit systems) or 32-bit integers. When using 64-bit integers, everything should work correctly. When using 32-bit integers, there should be a clear message about overflow, and no negative edge count.
